### DISCUSSION TITLE
different genes

### DISCUSSION THREAD

**Discussion by wbf520qcp** [campbio/celda#372] [Q&A] [answered]
Hi,
Thanks for developing this awesome tools. I have run it on my data and got fantastic figures. But I wonder how can I retrieve those gene modules so that I can manipulate heat maps with my own code?

Thank you,
Bofei

**Comment by joshua-d-campbell** [comment]
Hi @wbf520qcp, thanks for trying out our tools! Great question! If you want to get the gene modules, you can use the `factorizeMatrix` function described [here](https://www.camplab.net/celda/v1.12.0/articles/articles/celda_pbmc3k.html#other-useful-functions-1) in our tutorial. Briefly, if your celda result is stored in a *SingleCellExperiment* object called "sce", you can use the following code:

```
# Factorize the original counts matrix
fm <- factorizeMatrix(sce)

# Get the matrix of module probabilities for each cell
fm$proportions$cell

# Get the matrix of module counts for each cell
fm$counts$cell
```

Hope that helps. I'm going to convert this to a "Discussion".

**Comment by wbf520qcp** [comment]
Thanks for replying. This is helpful. I find out from the tutorial that featureModuleTable(sce, useAssay = useAssay, altExpName = altExpName, outputFile = "modules.txt") can output the genes in each module. I wonder if the those genes in each module ranked by fold change, p-value or just random? Because I want to select top differentially genes to plot Heatmap.

**Comment by joshua-d-campbell** [comment, marked as answer]
Good question. They are ranked by the estimated level (i.e. probability) of that gene in that particular module. This will generally correspond to the overall expression level, so genes ranked higher will have higher overall expression. You can get these ranks directly from the results of the `factorizeMatrix` function:

```
fm$proportions$module
``` 

In this matrix, the columns are modules and it shows the estimated (and normalized) level of each gene in each module. Each gene will only have a non-zero value in the module it was assigned to. 

You can also use the `moduleHeatmap` function to quickly plot a heatmap of the modules. Cells will be ordered from those with the lowest probability of that module to cells with the highest probability.
